httpbp/client_middlewares.go

```diff
@@ -0,0 +1,199 @@
+
+	retry "github.com/avast/retry-go"
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/reddit/baseplate.go/retrybp"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/sony/gobreaker"
+)
+
+// ClientMiddleware is used to build HTTP client middlewares by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(next http.RoundTripper) http.RoundTripper
+
+type roundTripper func(req *http.Request) (*http.Response, error)
+
+func (f roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middlewares plus any additional client middlewares passed into this
+// function.
+func NewClient(config ClientConfig, middlewares ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+
+	defaults := []ClientMiddleware{
+		MonitorClient(config),
+		Retries(),
+		CircuitBreaker(config),
+	}
+
+	middlewares = append(middlewares, defaults...)
+
+	return &http.Client{
+		Transport: applyMiddlewares(transport, middlewares...),
+	}
+}
+
+// ClientErrorWrapper applies ClientErrorFromResponse to the returned response
+// ensuring a HTTP status response outside the range [200, 400) is wrapped in
+// an error relieving users from the need to check the status response.
+//
+// Adding this middleware means that both, non-nil response and non-nil error,
+// can be returned. If configured the caller needs to call
+// httpbp.DrainAndClose(resp.Body) to ensure the underlying TCP connection can
+// be re-used.
+func ClientErrorWrapper() ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+			resp, err = next.RoundTrip(req)
+			err = ClientErrorFromResponse(resp)
+			return resp, err
+		})
+	}
+}
+
+// CircuitBreaker is a middleware that prevents sending requests that are likely
+// to fail based on a configurable failure ratio based on total failures and
+// requests.
+func CircuitBreaker(config ClientConfig) ClientMiddleware {
+	pool := &sync.Pool{
+		New: func() interface{} {
+			return &gobreaker.Settings{
+				Interval: 60 * time.Second,
+				ReadyToTrip: func(counts gobreaker.Counts) bool {
+					failureRatio := float64(counts.TotalFailures) / float64(counts.Requests)
+					return counts.Requests >= config.CircuitBreaker.MinRequests && failureRatio >= config.CircuitBreaker.MinFailureRatio
+				},
+			}
+		},
+	}
+	breakers := &sync.Map{}
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (*http.Response, error) {
+			host := req.URL.Hostname()
+
+			// new circuit breakers should rarely get allocated: the pool reduces
+			// garbage collector overhead
+			newBreaker := pool.Get()
+			breaker, loaded := breakers.LoadOrStore(host, newBreaker)
+			if loaded {
+				pool.Put(newBreaker)
+			}
```

Comment:
This seems dangerous... I assume that in this case `newBreaker == breaker`, right?  If so, then it could get pulled out and used concurrently.

Did you mean `defer pool.Put` or something?